internal/fourslash/test_parser.go

```diff
@@ -0,0 +1,258 @@
+)
+
+// Inserted in source files by surrounding desired text
+// in a range with `[|` and `|]`. For example,
+//
+// [|text in range|]
+//
+// is a range with `text in range` "selected".
+type markerRange struct {
+	core.TextRange
+	filename string
+	position int
+	data     string
+}
+
+type Marker struct {
+	Filename   string
+	Position   int
+	LSPosition lsproto.Position
+	Name       string
+}
+
+type TestData struct {
+	Files           []*TestFileInfo
+	MarkerPositions map[string]*Marker
+	Markers         []*Marker
+	Symlinks        map[string]string
+	GlobalOptions   map[string]string
+	Ranges          []*markerRange
+}
+
+type testFileWithMarkers struct {
+	file    *TestFileInfo
+	markers []*Marker
+}
+
+func ParseTestData(t *testing.T, contents string, fileName string) TestData {
+	// List of all the subfiles we've parsed out
+	var files []*TestFileInfo
+
+	markerPositions := make(map[string]*Marker)
+	var markers []*Marker
+	filesWithMarker, symlinks, _, globalOptions := runner.ParseTestFilesAndSymlinks(
+		contents,
+		fileName,
+		parseFileContent,
+	)
+
+	hasTSConfig := false
+	for _, file := range filesWithMarker {
+		files = append(files, file.file)
+		hasTSConfig = hasTSConfig || isConfigFile(file.file.Filename)
+		markers = append(markers, file.markers...)
+		for _, marker := range file.markers {
+			if _, ok := markerPositions[marker.Name]; ok {
+				t.Fatalf("Duplicate marker name: %s", marker.Name)
+			}
+			markerPositions[marker.Name] = marker
+		}
+	}
+
+	if hasTSConfig && len(globalOptions) > 0 {
+		t.Fatalf("It is not allowed to use global options along with config files.")
+	}
+
+	return TestData{
+		Files:           files,
+		MarkerPositions: markerPositions,
+		Markers:         markers,
+		Symlinks:        symlinks,
+		GlobalOptions:   globalOptions,
+		Ranges:          nil,
+	}
+}
+
+func isConfigFile(filename string) bool {
+	filename = strings.ToLower(filename)
+	return strings.HasSuffix(filename, "tsconfig.json") || strings.HasSuffix(filename, "jsconfig.json")
+}
+
+type locationInformation struct {
+	position       int
+	sourcePosition int
+	sourceLine     int
+	sourceColumn   int
+}
+
+type TestFileInfo struct {
+	Filename string
+	// The contents of the file (with markers, etc stripped out)
+	Content string
+	emit    bool
+}
+
+// FileName implements ls.Script.
+func (t *TestFileInfo) FileName() string {
+	return t.Filename
+}
+
+// Text implements ls.Script.
+func (t *TestFileInfo) Text() string {
+	return t.Content
+}
+
+var _ ls.Script = (*TestFileInfo)(nil)
+
+const emitThisFileOption = "emitthisfile"
+
+type parserState int
+
+const (
+	stateNone parserState = iota
+	stateInSlashStarMarker
+	stateInObjectMarker
+)
+
+func parseFileContent(filename string, content string, fileOptions map[string]string) *testFileWithMarkers {
+	filename = tspath.GetNormalizedAbsolutePath(filename, "/")
+	// Any slash-star comment with a character not in this string is not a marker.
+	validMarkerChars := core.NewSetFromItems(
```

Comment:
This seems like it could just be some calls to `stringutil`?

Reply:
What calls?

Reply:
I was just thinking `stringutil.IsASCIILetter(c) || stringutil.IsDigit(c) || c == '$' || c == '_'` or something